### PR TITLE
tscrypter/28/add asyncapi ui

### LINKED ISSUE #28: Add UI to view documentation
Similar to Swachbuckle.AspNetCore which provides `app.UseSwaggerUI()` it would be nice to have an `app.UseAsyncApiUI()`

### PR DESCRIPTION
**Pull request by tscrypter**
closes #28 
Everything is exposed via configuration. Defaults to making a request to `https://playground.asyncapi.io/`

### CODE REVIEW

**Comment by tscrypter** [comment]
Opened this up as a draft so you can play with it. I just need to add tests around some of the functionality in the middleware

**Comment by tehmantra** [comment]
Looks great! Have a few pointers & questions though. I'll add them as review comments

File: src/Saunter/AsyncApiEndpointRouteBuilderExtensions.cs [hunk 1]
```diff
@@ -17,12 +17,29 @@ public static class AsyncApiEndpointRouteBuilderExtensions
...
+            this IEndpointRouteBuilder endpoints)
+        {
+            // Add the middleware 
+            var pipeline = endpoints.CreateApplicationBuilder()
+                .UseMiddleware<AsyncApiUiMiddleware>()
+                .Build();
+
+            // Retrieve the route from options. If options can't be retrieved, use default constant
+            var options = endpoints.ServiceProvider.GetService<IOptions<AsyncApiOptions>>();
+            var route = "/asyncapi/ui/{resource}";
```
Comment (tehmantra) [inline src/Saunter/AsyncApiEndpointRouteBuilderExtensions.cs:37]: If you register a route like 
```csharp
var route = "/asyncapi/ui/{*resource}";
```
it is a catch-all for `/asyncapi/ui/any/number/of/segments`.

Which prevents you from having to register the 
```csharp
var subRoute = "/asyncapi/ui/{resource}/{subresource}";
```
route too.

File: src/Saunter/AsyncApiOptions.cs [hunk 2]
```diff
@@ -12,54 +11,101 @@ namespace Saunter
...
+        public string PlaygroundBaseAddress { get; set; } = AsyncApiPlaygroundBaseUri;
+
+        /// <summary>
+        ///     Key/Value pairs where keys will be replaced by the corresponding value in the html
+        ///     returned from playground
+        /// </summary>
+        public Dictionary<string, string> HtmlProxyRewrites { get; set; } = new Dictionary<string, string>
+        {
+            {
+                "text-xs uppercase text-grey mt-10 mb-4 font-thin",
```
Comment (tehmantra) [inline src/Saunter/AsyncApiOptions.cs:90]: Can you explain why these changes are made?

File: src/Saunter/AsyncApiUiMiddleware.cs [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+            {
+                content = content.Replace(key, value);
+            }
+            return content;
+        }
+
+        private bool IsRequestingAsyncApiUiAssets(HttpRequest request)
+        {
+            return HttpMethods.IsGet(request.Method) &&
+                   request.Path.ToString().ToLower().StartsWith(_options.Value.Middleware.UiBaseRoute);
```
Comment (tehmantra) [inline src/Saunter/AsyncApiUiMiddleware.cs:121]: Can use case-insensitive StartsWith here
```csharp
context.Path.ToString().StartsWith(_options.Value.Middleware.UiBaseRoute, StringComparison.OrdinalIgnoreCase);
```

File: src/Saunter/AsyncApiUiMiddleware.cs [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+            }
+
+            await RespondWithProxiedResponse(context.Request, context.Response);
+        }
+
+        private async Task RespondWithAsyncApiHtml(HttpResponse response, AsyncApiDocument asyncApiSchema)
+        {
+            if (!_memoryCache.TryGetValue<string>(PlaygroundHtmlCacheKey, out var responseString))
+            {
+                var asyncApiSchemaJson = JsonSerializer.Serialize(
```
Comment (tehmantra) [inline src/Saunter/AsyncApiUiMiddleware.cs:62]: Now we have 2 places which we need to get the serialized JSON (here and in AsyncApiMiddleware.cs), it would be a good idea to move this to the `IAsyncApiDocumentProvider` so we don't have to maintain this converter list in 2 places.

```csharp
public interface IAsyncApiDocumentProvider
{
    AsyncApiDocument GetDocument();

    string GetDocumentJson();
}
```

This `GetDocumentJson()` can call `GetDocument()` then `JsonSerializer.Serialize()` with the right options.

File: src/Saunter/Saunter.csproj [hunk 5]
```diff
@@ -41,4 +45,9 @@
     <FrameworkReference Include="Microsoft.AspNetCore.App" />
   </ItemGroup>
 
+  <ItemGroup>
```
Comment (tehmantra) [inline src/Saunter/Saunter.csproj:48]: These are no longer required.

File: src/Saunter/Saunter.csproj [hunk 6]
```diff
@@ -23,6 +23,10 @@
     <ContinuousIntegrationBuild>true</ContinuousIntegrationBuild>
   </PropertyGroup>
 
+  <ItemGroup>
```
Comment (tehmantra) [inline src/Saunter/Saunter.csproj:26]: This is no longer required.

**Review by tehmantra — CHANGES_REQUESTED** [review]

File: src/Saunter/AsyncApiEndpointRouteBuilderExtensions.cs (hunk 1 above)
Comment (tscrypter) [reply src/Saunter/AsyncApiEndpointRouteBuilderExtensions.cs:37]: I'll make that update.

File: src/Saunter/AsyncApiOptions.cs (hunk 2 above)
Comment (tscrypter) [reply src/Saunter/AsyncApiOptions.cs:90]: The styles being served for some of the content appears to be styled to show up in playground where there is an additional style sheet that sets some darker backgrounds. This results in the html generated by the generator serving some light-grey font on a light background.
I made use of this functionality to rewrite some of the styling to fix that, but ultimately I could see this being more formalized to enable some theme-ing of the AsyncApiUI.
